ledger/accountdb.go

```diff
@@ -192,6 +203,156 @@ func prepareNormalizedBalances(bals []encodedBalanceRecord, proto config.Consens
 	return
 }
 
+// makeCompactAccountDeltas takes an array of account AccountDeltas ( one array entry per round ), and compacts the arrays into a single
+// data structure that contains all the account deltas changes. While doing that, the function eliminate any intermediate account changes.
+// It counts the number of changes per round by specifying it in the ndeltas field of the accountDeltaCount/modifiedCreatable.
+func makeCompactAccountDeltas(accountDeltas []ledgercore.AccountDeltas, baseAccounts lruAccounts) (outAccountDeltas compactAccountDeltas) {
+	if len(accountDeltas) == 0 {
+		return
+	}
+
+	// the sizes of the maps here aren't super accurate, but would hopefully be a rough estimate for a reasonable starting point.
+	size := accountDeltas[0].Len()*len(accountDeltas) + 1
+	outAccountDeltas.cache = make(map[basics.Address]int, size)
+	outAccountDeltas.deltas = make([]accountDelta, 0, size)
+	outAccountDeltas.misses = make([]int, 0, size)
+
+	for _, roundDelta := range accountDeltas {
+		for i := 0; i < roundDelta.Len(); i++ {
+			addr, acctDelta := roundDelta.GetByIdx(i)
+			if prev, idx := outAccountDeltas.get(addr); idx != -1 {
+				outAccountDeltas.update(idx, accountDelta{ // update instead of upsert economizes one map lookup
+					old:     prev.old,
+					new:     acctDelta,
+					ndeltas: prev.ndeltas + 1,
+				})
+			} else {
+				// it's a new entry.
+				newEntry := accountDelta{
+					new:     acctDelta,
+					ndeltas: 1,
+				}
+				if baseAccountData, has := baseAccounts.read(addr); has {
+					newEntry.old = baseAccountData
+					outAccountDeltas.insert(addr, newEntry) // insert instead of upsert economizes one map lookup
+				} else {
+					outAccountDeltas.insertMissing(addr, newEntry)
+				}
+			}
+		}
+	}
+	return
+}
+
+// accountsLoadOld updates the entries on the deltas.old map that matches the provided addresses.
+// The round number of the persistedAccountData is not updated by this function, and the caller is responsible
+// for populating this field.
+func (a *compactAccountDeltas) accountsLoadOld(tx *sql.Tx) (err error) {
```

Comment:
how's about updating the `compactAccountDeltas.misses` so that it would be set to nil once you exit from this function ? ( i.e. since that array is no longer needed )

Reply:
done